proposals/0034-glulx-random.md

```diff
@@ -0,0 +1,134 @@
+interpreters have since been patched to use much more reliable RNGs, but it can
+take a long time for downstream interpreters to include these updates; many
+interpreters are only updated infrequently, and some are no longer maintained,
+including all Android Glulx interpreters.
+
+Inform must provide high quality randomness out-of-the-box without authors needing
+to dig into interpreter quirks. But as we can't rely on the interpreters being
+patched in a timely manner, we need to instead move the implementation of Inform's
+RNG inside Inform kit code (Architecture32Kit) to insulate authors from the flaws
+of the interpreters. Many other programming languages also implement their RNGs
+inside their library code rather than relying on the OS-provided RNG, so Inform
+will be in good company doing this.
+
+## Components affected
+
+- [ ] No change to the natural-language syntax.
+- [ ] No change to inbuild.
+- [x] Minor changes to inform7.
+- [ ] No change to inter.
+- [ ] No change to the Inter specification.
+- [x] Minor changes to runtime kits.
+- [ ] No changes to the Standard Rules and Basic Inform.
+- [ ] No change to documentation.
+- [ ] No change to the GUI apps.
+
+## Impact on existing projects
+
+Authors will receive better quality random numbers without any change to their
+code. However if anyone is unintentionally relying on the quirky results of
+calling `random` with a negative value in Glulx, they will need to change what
+they are doing.
+
+## Details
+
+The Inform 6 function `random` is a "system" function. This means that by default
+it is not a true function, but instead just a shortcut for the VM-specific random
+opcode. System functions can however be replaced by the author, so that is what
+we will do to fix this problem, replacing it with a function in Architecture32Kit.
+
+Considerations:
+
+1. Ideally we want an RNG that produces high quality random numbers, has decently
+    high performance, and does not use excessive amounts of memory. The [Xorshift](https://en.wikipedia.org/wiki/Xorshift)
+    family of RNGs fits these criteria. Glulxe and Git switched to the
+    [xoshiro128** algorithm](https://prng.di.unimi.it/), so we might as well
+    use it too.
+
+2. This RNG can be seeded with any non-zero value, and it is common to seed it
+    with a timestamp. The Inform test suite however requires predictable randomness,
+    which it currently accomplishes through Glulxe's `--rngseed` argument. So
+    rather than seeding Inform's new RNG with a timestamp, we should instead seed
+    it with `@random`. Even though the interpreter's RNG may be flawed, one call
+    with a high enough range is probably much safer than calling it with a range
+    of only 4 or 8. It has not yet been determined whether a full 32-bit `@random`
+    call is safe in affected interpreters, or instead whether it should make a
+    ranged call using a high prime number, or if any priming is necessary.

```

Comment:
Hmm good point. I was intending to make the new RNG avoid rounding biases, but we can't assume the interpreter RNG is doing that. We could get a full 32bit number and do it ourselves if necessary (or just use the 32 but result as the seed if that's safe enough).

Reply:
Looks like [Git before version 1.3.8](https://github.com/DavidKinder/Git/commit/1f24f3e9b7f71d2ab7344b2405e5b43f8aaf1d4e) always used the C standard library's `rand()` function, whose range and quality is very platform-dependent. It's especially limited on Windows (0x7fff), but since it returns a non-negative `int`, it can be at most 31 bits on any platform with 32-bit `int` (which includes all common 64-bit platforms). Together with the *other* RNG flaw we know to be common, very low quality in the lowest bits, this makes it very annoying to reliably extract a decent 32-bit number. I'm sure there are decent ways, but I'll have to think about it some more.

Reply:
That's informative, thanks. But unless you think it will be impossible to construct a useful seed (possibly from many calls to `@random`, this level of detailed analysis would probably be more helpful in the forum discussion than in this proposal pull request. The proposal doesn't specify all implementation details, just the broad plan. 🙂